File: leet_code_solutions.js/16.js
// Solved.
// Runtime: better than 99.83% of JS solutions
// Memory Usage: better than 30.15% of JS solutions

/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
var threeSumClosest = function (nums, target) {
  nums.sort((a, b) => a - b);
  var minDiff = 1000;
  var sum;
  for (let i = 0; i < nums.length; i++) {
    let start = i + 1;
    let end = nums.length - 1;
    while (start < end) {
      sum = nums[i] + nums[start] + nums[end];
      if (Math.abs(minDiff) > Math.abs(target - sum)) {
        minDiff = target - sum;
      }
      if (sum < target) {
        start++;
      } else {
        end--;
      }
      if (minDiff === 0) {
        break;
      }
    }
  }
  return target - minDiff;
};

console.log(threeSumClosest([-1, 2, 1, -4], 1));

File: leet_code_solutions.py/16.py

def threeSumClosest(nums: List[int], target: int) -> int:

        nums.sort()
        Min=10**4

        for i in range(len(nums)-1):
            low=i+1
            high=len(nums)-1

            while(low<high):
                diff = abs(nums[i]+nums[low]+nums[high]-target)

                if diff==0:
                    return nums[i]+nums[low]+nums[high]

                if diff<=Min:
                    ans = nums[i] + nums[low] + nums[high]
                    Min = diff

                if nums[i]+nums[low]+nums[high]<=target:
                    low+=1
                else:
                    high-=1
        return ans
print(threeSumClosest([[-1,2,1,-4]))
